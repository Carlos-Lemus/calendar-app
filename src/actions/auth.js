import Swal from "sweetalert2";
import { fetchWithoutToken, fetchWithToken } from "../helpers/fetch";
import { types } from "../types/types";

export const startLogin = (email, password) => {
    return async(dispatch) => {
        
        const req = await fetchWithoutToken("auth", "POST", {email, password});
        const body = await req.json();

        if(body.ok) {

            localStorage.setItem("token", body.token);
            localStorage.setItem("token-init-date", new Date().getTime());

            dispatch(login({
                uid: body.uid,
                name: body.name
            }));
        } else {
            Swal.fire("Error", body.msg, "error");
        }

    };
};

export const startRegister = (name, email, password) => {
    return async(dispatch) => {
        const resp = await fetchWithoutToken("auth/new", "POST", {name, email, password});
        const body = await resp.json();
        
        if(body.ok) {
            localStorage.setItem("token", body.token);
            localStorage.setItem("token-init-date", new Date().getTime());
            
            dispatch(login( { name: body.name, uid: body.uid } ) );
            
        } else {
            Swal.fire("Error", body.msg, "error");
        }
    };
};

export const startChecking = () => {
    return async(dispatch) => {
        const resp = await fetchWithToken("auth/renew");
        const body = await resp.json();
        
        if(body.ok) {
            localStorage.setItem("token", body.token);
            localStorage.setItem("token-init-date", new Date().getTime());
            
            dispatch(login( { name: body.name, uid: body.uid } ) );
            
        } else {
            dispatch(finishChecking());
        }
    };
};

export const startLogout = () => {
    return async(dispatch) => {

        localStorage.clear();

        dispatch(logout());
    }
};

const logout = () => ({ type: types.authLogout });

const login = ( user ) => ({
    type: types.authLogin,
    payload: user
});

export const finishChecking = () => ({
    type: types.authCheckingFinish
});

