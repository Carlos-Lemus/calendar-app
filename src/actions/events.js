import Swal from "sweetalert2";
import { fetchWithToken } from "../helpers/fetch";
import { prepareEvents } from "../helpers/prepareEvents";
import { types } from "../types/types";

export const startLoadingEvents = () => {
    return async(dispatch) => {
        try {
            const res = await fetchWithToken("events/");
            const body = await res.json();
            
            const events = prepareEvents(body.events);

            dispatch(loadedEvents(events));
            
        } catch (error) {
            console.log(error)
        }
    }
};

const loadedEvents = (events) => ({
    type: types.eventLoaded,
    payload: events
});

export const startSetActiveEvent = (event) => ({
    type: types.eventSetActive,
    payload: event
});

export const startNewAddEvent = (event) => {
    return async (dispatch, getState) => {
        try {
            const res = await fetchWithToken("events", "POST", event);
            const body = await res.json();

            if (body.ok) {

                const { uid, name } = getState().auth;

                event.id = body.event.id;
                event.user = {
                    _id: uid,
                    name
                };

                dispatch(newAddEvent(event));
            } else {
                Swal.fire("Error", body.msg, "error");
            }
        } catch (error) {
            console.log(error)
        }
    };
};

const newAddEvent = (event) => ({
    type: types.eventAddNew,
    payload: event
});

export const startClearActiveEvent = () => ({
    type: types.eventClearActiveEvent
});

export const startUpdateEvent = (event) => {
    return async(dispatch) => {
        try {
            const res = await fetchWithToken(`events/${ event.id }`, "PUT", event);
            const body = await res.json();
            
            if(body.ok) {
                dispatch(updateEvent(event));
            } else {
                Swal.fire("Error", body.msg, "error");
            }

        } catch (error) {
            console.log(error);
        }
    };
};

const updateEvent = (event) => ({
    type: types.eventUpdate,
    payload: event
});

export const startDeleteEvent = () => {

    return async(dispatch, getState) => {

        try {
            const {id} = getState().calendar.activeEvent;

            const res = await fetchWithToken(`events/${id}`, "DELETE");
            const body = await res.json();

            if(body.ok) {

                dispatch(deleteEvent());       
                dispatch(startClearActiveEvent());

            } else {
                Swal.fire("Error", body.msg, "error");
            }

        } catch (error) {
            console.log(error)
        }
    };

};

const deleteEvent = () => ({
    type: types.eventDeleted
});