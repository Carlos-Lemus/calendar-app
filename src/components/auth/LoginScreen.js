import React from 'react';
import { useDispatch } from 'react-redux';
import Swal from 'sweetalert2';
import { startLogin, startRegister } from '../../actions/auth';
import { useForm } from '../../hooks/useForm';
import './login.css';

export const LoginScreen = () => {
    
    const dispatch = useDispatch();

    const [formLoginValues, changeInputFormLoginValues] = useForm({
        lEmail: "pedro@gmail.com",
        lPassword: "123456"
    });

    const { lEmail, lPassword } = formLoginValues;
    

    const handleLogin = (event) => {
        event.preventDefault();
        
        dispatch(startLogin(lEmail, lPassword));
    }

    const [formRegisterValues, changeInputFormRegisterValues] = useForm({
        rName: "Pedro",
        rEmail: "pedro2@gmail.com",
        rPassword1: "123456",
        rPassword2: "123456"
    });

    const { rName, rEmail, rPassword1, rPassword2 } = formRegisterValues;
    
    const handleRegister = (event) => {
        event.preventDefault();

        if(rPassword1 === rPassword2 ){
            dispatch(startRegister(rName, rEmail, rPassword1));
        } else {
            Swal.fire("Error", "Los password no son iguales", "error");
        }

    }

    return (
        <div 
            style={ 
                { 
                    backgroundImage: "url(/assets/img/background.jpg)",
                    backgroundSize: "cover",
                    backgroundRepeat: "no-repeat",
                    backgroundPosition: "left bottom"
                } 
            }
            className="container-fluid login-container"
        >
            <div className="row">
                <div className="col-md-6 login-form-1">
                    <h3>Ingreso</h3>
                    <form onSubmit={ handleLogin } >
                        <div className="form-group">
                            <input
                                name="lEmail"
                                onChange={ changeInputFormLoginValues } 
                                value={ lEmail }
                                type="text"
                                className="form-control"
                                placeholder="Correo"
                            />
                        </div>
                        <div className="form-group">
                            <input
                                name="lPassword"
                                onChange={ changeInputFormLoginValues }
                                value={ lPassword }
                                type="password"
                                className="form-control"
                                placeholder="Contraseña"
                            />
                        </div>
                        <div className="form-group">
                            <input 
                                type="submit"
                                className="btnSubmit"
                                value="Login" 
                            />
                        </div>
                    </form>
                </div>

                <div className="col-md-6 login-form-2">
                    <h3>Registro</h3>
                    <form onSubmit={ handleRegister } >
                        <div className="form-group">
                            <input
                                name="rName"
                                value={rName}
                                onChange={ changeInputFormRegisterValues }
                                type="text"
                                className="form-control"
                                placeholder="Nombre"
                            />
                        </div>
                        <div className="form-group">
                            <input
                                name="rEmail"
                                value={rEmail}
                                onChange={ changeInputFormRegisterValues }
                                type="email"
                                className="form-control"
                                placeholder="Correo"
                            />
                        </div>
                        <div className="form-group">
                            <input
                                name="rPassword1"
                                value={rPassword1}
                                onChange={ changeInputFormRegisterValues }
                                type="password"
                                className="form-control"
                                placeholder="Contraseña" 
                            />
                        </div>

                        <div className="form-group">
                            <input
                                name="rPassword2"
                                value={rPassword2}
                                onChange={ changeInputFormRegisterValues }
                                type="password"
                                className="form-control"
                                placeholder="Repita la contraseña" 
                            />
                        </div>

                        <div className="form-group">
                            <input 
                                type="submit" 
                                className="btnSubmit" 
                                value="Crear cuenta" />
                        </div>
                    </form>
                </div>
            </div>
        </div>
    )
}