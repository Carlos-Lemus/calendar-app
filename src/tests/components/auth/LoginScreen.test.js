import configMockStore from "redux-mock-store";
import thunk from "redux-thunk";
import "@testing-library/jest-dom";
import { mount } from "enzyme";
import { Provider } from "react-redux";
import { LoginScreen } from "../../../components/auth/LoginScreen";
import { startLogin, startRegister } from "../../../actions/auth";
import Swal from "sweetalert2";

const middlewares = [thunk];
const mockStore = configMockStore(middlewares);

const initialState = {  };

let store = mockStore(initialState);
store.dispatch = jest.fn();

jest.mock("../../../actions/auth", () => ({
    startLogin: jest.fn(),
    startRegister: jest.fn()
}));

jest.mock("sweetalert2", () => ({
    fire: jest.fn()
}));

const wrapper = mount(
    <Provider store={store}>
        <LoginScreen />
    </Provider>
);

describe("Pruebas en <LoginScreen />", () => {

    beforeEach(() => {
        jest.clearAllMocks();
    });

    test('<LoginScreen /> debe de mostrarse correctamente', () => {
        expect(wrapper).toMatchSnapshot(); 
    });
    
    test('<LoginScreen /> debe de hacer el login correctamente', () => {
        
        wrapper.find("input[name='lEmail']").simulate("change", {
            target: {
                name: "lEmail",
                value: "pedro@gmail.com"
            }
        });

        wrapper.find("input[name='lPassword']").simulate("change", {
            target: {
                name: "lPassword",
                value: "123456"
            }
        });

        wrapper.find("form").at(0).prop("onSubmit") ({
            preventDefault() {}
        });

        expect(startLogin).toHaveBeenCalledWith("pedro@gmail.com", "123456");
    });
   
    test('No hay registros si la contraseÃ±a es diferente', () => {
        wrapper.find("input[name='rPassword1']").simulate("change", {
            target: {
                name: "rPassword1",
                value: "123456"
            }
        });

        wrapper.find("input[name='rPassword2']").simulate("change",{
            target: {
                name: "rPassword2",
                value: "12345678"
            }
        });

        wrapper.find("form").at(1).prop("onSubmit")({
            preventDefault() {}
        });

        expect(startRegister).not.toHaveBeenCalled();
        expect(Swal.fire).toHaveBeenCalledTimes(1);
        expect(Swal.fire).toHaveBeenCalledWith("Error", "Los password no son iguales", "error");
    });

    test('Si los datos son correctos se registra', () => {
        wrapper.find("input[name='rEmail']").simulate("change", {
            target: {
                name: "rEmail",
                value: "test23@gmail.com"
            }
        });

        wrapper.find("input[name='rPassword1']").simulate("change", {
            target: {
                name: "rPassword1",
                value: "123456"
            }
        });

        wrapper.find("input[name='rPassword2']").simulate("change",{
            target: {
                name: "rPassword2",
                value: "123456"
            }
        });

        wrapper.find("form").at(1).prop("onSubmit")({
            preventDefault() {}
        });

        expect(startRegister).toHaveBeenCalledWith("Pedro", "test23@gmail.com", "123456");
        expect(Swal.fire).toHaveBeenCalledTimes(0);
        expect(Swal.fire).not.toHaveBeenCalled();
    });
    

});