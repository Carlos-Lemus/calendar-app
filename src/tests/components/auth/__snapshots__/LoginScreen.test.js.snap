// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pruebas en <LoginScreen /> <LoginScreen /> debe de mostrarse correctamente 1`] = `
<div
  className="container-fluid login-container"
  style={
    Object {
      "backgroundImage": "url(/assets/img/background.jpg)",
      "backgroundPosition": "left bottom",
      "backgroundRepeat": "no-repeat",
      "backgroundSize": "cover",
    }
  }
>
  <div
    className="row animate__animated animate__fadeIn"
  >
    <div
      className="col-md-6 login-form-1"
    >
      <h3>
        Ingreso
      </h3>
      <form
        onSubmit={[Function]}
      >
        <div
          className="form-group"
        >
          <input
            className="form-control"
            name="lEmail"
            onChange={[Function]}
            placeholder="Correo"
            type="text"
            value="pedro@gmail.com"
          />
        </div>
        <div
          className="form-group"
        >
          <input
            className="form-control"
            name="lPassword"
            onChange={[Function]}
            placeholder="Contraseña"
            type="password"
            value="123456"
          />
        </div>
        <div
          className="form-group"
        >
          <input
            className="btnSubmit"
            type="submit"
            value="Login"
          />
        </div>
      </form>
    </div>
    <div
      className="col-md-6 login-form-2"
    >
      <h3>
        Registro
      </h3>
      <form
        onSubmit={[Function]}
      >
        <div
          className="form-group"
        >
          <input
            className="form-control"
            name="rName"
            onChange={[Function]}
            placeholder="Nombre"
            type="text"
            value="Pedro"
          />
        </div>
        <div
          className="form-group"
        >
          <input
            className="form-control"
            name="rEmail"
            onChange={[Function]}
            placeholder="Correo"
            type="email"
            value="pedro2@gmail.com"
          />
        </div>
        <div
          className="form-group"
        >
          <input
            className="form-control"
            name="rPassword1"
            onChange={[Function]}
            placeholder="Contraseña"
            type="password"
            value="123456"
          />
        </div>
        <div
          className="form-group"
        >
          <input
            className="form-control"
            name="rPassword2"
            onChange={[Function]}
            placeholder="Repita la contraseña"
            type="password"
            value="123456"
          />
        </div>
        <div
          className="form-group"
        >
          <input
            className="btnSubmit"
            type="submit"
            value="Crear cuenta"
          />
        </div>
      </form>
    </div>
  </div>
</div>
`;
